translator/reader/reader.cpp

```diff
@@ -158,25 +159,74 @@ unsigned int UsdArnoldReader::RenderThread(void *data)
     TfToken visibility;
     const TimeSettings &time = reader->GetTimeSettings();
     float frame = time.frame;
-
+    // Each thread context will have a stack of primvars vectors,
+    // which represent the primvars at the current level of hierarchy.
+    // Every time we find a Xform prim, we add an element to the stack 
+    // with the updated primvars list. In every "post" visit, we pop the last
+    // element. Thus, every time we'll read a prim, the last element of this 
+    // stack will represent its input primvars that it inherits (see #282)
+    std::vector<std::vector<UsdGeomPrimvar> > &primvarsStack = threadData->context.GetPrimvarsStack();
+    primvarsStack.clear(); 
+    primvarsStack.reserve(64); // reserve first to avoid frequent memory allocations
+    primvarsStack.push_back(std::vector<UsdGeomPrimvar>()); // add an empty element first
+    
     // Traverse the stage, either the full one, or starting from a root primitive
-    // (in case an object_path is set).
-    UsdPrimRange range = (rootPrim) ? UsdPrimRange(*rootPrim) : reader->GetStage()->Traverse();
+    // (in case an object_path is set). We need to have "pre" and "post" visits in order
+    // to keep track of the primvars list at every point in the hierarchy.
+    UsdPrimRange range = UsdPrimRange::PreAndPostVisit((rootPrim) ? 
+                    *rootPrim : reader->GetStage()->GetPseudoRoot());
     for (auto iter = range.begin(); iter != range.end(); ++iter) {
         const UsdPrim &prim(*iter);
-        // Check if that primitive is set as being invisible.
-        // If so, skip it and prune its children to avoid useless conversions
 
-        // Special case for arnold schemas, they don't inherit from UsdGeomImageable
-        // but we author these attributes nevertheless
         std::string objType = prim.GetTypeName().GetText();
-        if (prim.IsA<UsdGeomImageable>() || objType.substr(0, 6) == "Arnold") {
+        // skip untyped primitives
+        if (objType.empty())
+            continue;
+
+        // We traverse every primitive twice : once from root to leaf, 
+        // then back from leaf to root. We don't want to anything during "post" visits
+        // apart from popping the last element in the primvars stack for xforms.
+        // This way, the last element in the stack will always match the current 
+        // set of primvars
+        if (iter.IsPostVisit()) {
+            if (objType == "Xform")
```

Comment:
I don't think checking if objType is "Xform" is the right approach here, UsdGeomPRimvarsAPI can apply to almost anything, not just Xforms. For example, UsdGeom also defines "Scope" which can also hold primvars and can be used to group objects together. Also, meshes can have other meshes embedded, USD allows for a fully flexible scenegraph.

I think it's easier just to initialize UsdGeomPrimvarsAPI and check if it's valid. (ie cast it to a bool, like `if (primvarsAPI) ...` )

Reply:
yeah I know.... what I was trying to do here is to only compute this uselessly for non-leaf primitives.
The docs say 
```
" It is not generally useful to call this method on UsdGeomGprim leaf
    /// prims, and furthermore likely to be expensive since *most* primvars
    /// are defined on Gprims."
```
but I haven't found an efficient way to know if a prim has children or not (I need to call it on all prims during traversal).

That being said.... for leaf primitives, we'll usually call the same function again in `_ReadArnoldParameters` to read the `primvars:arnold` attributes. So another approach could be to always do it in the loop, and then in `_ReadArnoldParameters` just read the previously computed set of primvars. I'll try that

Reply:
Pushed the change. We now compute the primvars list for every prim type during the traversal.
Then in `ReadArnoldParameters`  (which is called for nearly all primitives) I just reuse this last element of the stack instead of recomputing it.